### PR TITLE
Expose the Python API for RuleGraph construction

### PR DESCRIPTION
**Pull request by ketkarameya**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #369
* #364
* #363
* #362

### CODE REVIEW

File: tests/tests.py [hunk 1]
```diff
@@ -67,6 +66,83 @@ def test_piranha_match_only():
...
+            (class_declaration name: (_)@class_name 
+                body : (class_body ((_)*) @class_members)  @class_body
+            ) @class_declaration
+            (#eq? @class_name "FooBar")
+        )
+        """,
+        replace_node="class_body",
+        replace="""{
+  private List<String> names;\n @class_members
+}""",
```
Comment (lazaroclapp) [inline tests/tests.py:84]: I guess one interesting caveat of the python API is that multi-line strings force us to have some somewhat awkward indentation here in order to get the right indentation in the final file. Finally one point in favor of the otherwise annoying EP varargs `lines` approach (see [here](https://github.com/uber/NullAway/blob/master/nullaway/src/test/java/com/uber/nullaway/NullAwayCoreTests.java#L74) for example)

File: tests/tests.py [hunk 2]
```diff
@@ -67,6 +66,83 @@ def test_piranha_match_only():
...
+        )
+
+    args = PiranhaArguments(
+        "test-resources/java/insert_field_and_import/input",
+        "test-resources/java/insert_field_and_initializer/configurations",
+        "java",
+        {
+            "stale_flag_name": "STALE_FLAG",
+            "treated": "true",
+            "treated_complement": "false",
```
Comment (lazaroclapp) [inline tests/tests.py:135]: Why do we need any sort of configuration about stale flags to run the field insertion example above? Seems unrelated to flags.

File: src/models/constraint.rs [hunk 3]
```diff
@@ -11,25 +11,40 @@ Copyright (c) 2022 Uber Technologies, Inc.
  limitations under the License.
 */
 
+use super::default_configs::{default_matcher, default_queries};
 use derive_builder::Builder;
 use getset::Getters;
+use pyo3::prelude::*;
```
Comment (lazaroclapp) [inline src/models/constraint.rs:17]: Not sure about Rust best practices, but for every other language our coding guidelines tend to be _against_ star imports.

File: src/models/outgoing_edges.rs [hunk 4]
```diff
@@ -23,22 +23,43 @@ pub(crate) struct Edges {
 
 // Captures an entry from the `edges.toml` file.
 #[derive(Deserialize, Debug, Clone, Hash, PartialEq, Eq, Default, Getters, Builder)]
+#[pyclass]
 pub struct OutgoingEdges {
+  /// The source rule or group of rules
   #[get = "pub with_prefix"]
-  from: String,
+  #[serde(alias = "from")]
```
Comment (lazaroclapp) [inline src/models/outgoing_edges.rs:30]: Interesting. This is due to `from` being a keyword in Python, right?

**Review by lazaroclapp — COMMENTED** [review]
Some comments below, but overall makes sense to me.

**Comment by ketkarameya** [comment]
Ow. I merged this by mistake. I reverted the commit( on master). I will make PR again by cherrypicking the commit.

File: src/models/outgoing_edges.rs (hunk 4 above)
Comment (ketkarameya) [reply src/models/outgoing_edges.rs:30]: Yes. :O
